tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml

```diff
@@ -0,0 +1,282 @@
+
+  manufacturer_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_MANUFACTURER
+  manufacturer_length:
+    _type: uint16_t
+    _value_const: WSC_MAX_NAME_LENGTH
+    #_length_var: true
+  manufacturer:
+    _type: char
+    _length: [ WSC_MAX_NAME_LENGTH ]
+
+  model_name_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_MODEL_NAME
+  model_name_length:
+    _type: uint16_t
+    _value_const: WSC_MAX_NAME_LENGTH
+    #_length_var: true
+  model_name:
+    _type: char
+    _length: [ WSC_MAX_NAME_LENGTH ]
+
+  model_number_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_MODEL_NUMBER
+  model_number_length:
+    _type: uint16_t
+    _value_const: WSC_MAX_NAME_LENGTH
+    #_length_var: true
+  model_number:
+    _type: char
+    _length: [ WSC_MAX_NAME_LENGTH ]
+
+  serial_number_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_SERIAL_NUMBER
+  serial_number_length:
+    _type: uint16_t
+    _value_const: WSC_MAX_NAME_LENGTH
+    #_length_var: true
+  serial_number:
+    _type: char
+    _length: [ WSC_MAX_NAME_LENGTH ]
+
+  primary_device_type_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_PRIMARY_DEV_TYPE
+  primary_device_type_length:
+    _type: uint16_t
+    _value_const: WSC_PRIMARY_DEV_TYPE_LENGTH
+  primary_device_type_category_id:
+    _type: uint16_t 
+  primary_device_type_oui:
+    _type: uint8_t
+    _length: [ WSC_PRIMARY_DEV_TYPE_OUI_LENGTH ]
+  primary_device_type_subcategory_id:
+    _type: uint16_t 
+
+  rf_bands_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_RF_BANDS
+  rf_bands_length:
+    _type: uint16_t
+    _value_const: 1
+  rf_bands:
+    _type: eWscValues8
+
+  association_state_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_ASSOC_STATE
+  association_state_length:
+    _type: uint16_t
+    _value_const: 1
+  association_state:
+    _type: eWscValues8
+    _value_const: WSC_ASSOC_CONN_SUCCESS
+
+  configuration_error_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_CONFIG_ERROR
+  configuration_error_length:
+    _type: uint16_t
+    _value_const: 2
+  configuration_error:
+    _type: eWscValues16
+    _value_const: WSC_CFG_NO_ERROR
+
+  device_password_id_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_DEV_PASSWORD_ID
+  device_password_id_length:
+    _type: uint16_t
+    _value_const: 2
+  device_password_id:
+    _type: eWscValues16
+    _value_const: DEV_PW_PUSHBUTTON
+
+  os_version_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_OS_VERSION
+  os_version_length:
+    _type: uint16_t
+    _value_const: WSC_OS_VERSION_LENGTH
+  os_version:
+    _type: uint32_t
+    _value_const: 0x80000001
+
+  vendor_extensions_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_VENDOR_EXTENSION
+  vendor_extensions_length:
+    _type: uint16_t
+    _value_const: WSC_VENDOR_EXTENSIONS_LENGTH
+  vendor_extensions:
+    _type: uint8_t
+    _length: [ WSC_VENDOR_EXTENSIONS_LENGTH ]
+
+  #encrypted data
```

Comment:
The encrypted data should be wrapped in a KEY_WRAP_AUTH attribute. And in hostapd/wpa_supplicant, it is also wrapped in an inner CRED attribute, though the BBF implementation of 1905.1 doesn't do the latter.

Reply:
I was mistaken, it's wrapped in an ENCRYPTED_SETTINGS attribute, and the last (nested) attribute is KEY_WRAP_AUTH.

So I have four sources for the WSC stuff (and all of them are inconsistent :-)

1. The Multi-AP specification. It doesn't say much about how M1 and M2 have to be constructed, so not much help from there.
2. The WSC specification. This is a huge document and the information is spread all over hte place, so difficult to draw solid conclusions from it. For some reason, the M2 description (v2.0.5, section 8.3.2, p.75) doesn't mention encrypted settings, but later on tables 20 and 21 in section 8.3.9 (M8) do specify that they apply to M2 as well. The "Encrypted Settings" part in section 12 explains how this attribute is constructed, though it is a bit vague as to whether the length is just for the IV or also includes the following (encrypted) attributes. Back to Table 21, it specifies "Credential" as a required attribute, and in section 12, the Credential attribute is specified to contain SSID etc. But now I read the spec again, I see that for the M2 used in AP-Autoconfig, we have to use Table 20 instead of Table 21, and that one doesn't have "Credential", it just has SSID etc. directly there. The description of the Encrypted Settings attribute also specifies that it must end with a Key Wrap Authenticator attribute.
3. The hostapd implementation. Reading the source code, hostapd always assumes that the settings are in the Encrypted Settings attribute, and that the SSID etc. are wrapped into an inner Credential attribute. So it always follows Table 21, AFAICS.
4. The BBF meshComms implementation. This one also has the Encrypted Settings attribute, but doesn't nest into a Credential attribute.
 
Based on all of the above, my conclusion is that we need an Encrypted Settings attribute that contains (cfr. Table 20) type, length (of the entire thing), SSID, AUTH_TYPE, ENCR_TYPE, NETWORK_KEY, BSSID, KEY_WRAP_AUTH. This is exactly how meshComms constructs it. Since hostapd is never involved in this specifc M2, it's not a problem that hostapd does it differently.

Reply:
Thank you very much!
The specs are indeed a bit confusing, but you summed it up very well.
I uploaded a PR for the refactoring (in the unified repo) and added this change in a fixup commit.